src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -546,40 +546,48 @@ void ShenandoahOldHeuristics::clear_triggers() {
-    //   else if old_gen spans more than 62.5% and density < 37.5%, trigger old-defrag
-    //   else if old_gen spans more than 50% and density < 25%, trigger old-defrag
-    //
-    // A previous implementation was more aggressive in triggering, resulting in degraded throughput when
-    // humongous allocation was not required.
-
-    size_t old_available = _old_gen->available();
-    size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
-    size_t old_unaffiliated_available = _old_gen->free_unaffiliated_regions() * region_size_bytes;
+    size_t old_available = _old_gen->available() / HeapWordSize;
+    size_t region_size_words = ShenandoahHeapRegion::region_size_words();
+    size_t old_unaffiliated_available = _old_gen->free_unaffiliated_regions() * region_size_words;
     assert(old_available >= old_unaffiliated_available, "sanity");
     size_t old_fragmented_available = old_available - old_unaffiliated_available;
 
-    size_t old_bytes_consumed = old_region_count * region_size_bytes - old_fragmented_available;
-    size_t old_bytes_spanned = old_region_span * region_size_bytes;
-    double old_density = ((double) old_bytes_consumed) / old_bytes_spanned;
-
-    uint eighths = 8;
-    for (uint i = 0; i < 5; i++) {
-      size_t span_threshold = eighths * allowed_old_gen_span / 8;
-      double density_threshold = (eighths - 2) / 8.0;
-      if ((old_region_span >= span_threshold) && (old_density < density_threshold)) {
-        trigger_old_is_fragmented(old_density, first_old_region, last_old_region);
-        return;
-      }
-      eighths--;
+    size_t old_words_consumed = old_region_count * region_size_words - old_fragmented_available;
+    size_t old_words_spanned = old_region_span * region_size_words;
+    double old_density = ((double) old_words_consumed) / old_words_spanned;
+
+    double old_span_percent = ((double) old_region_span) / allowed_old_gen_span;
+    double old_span_percent_squared = old_span_percent * old_span_percent;
+
+    // Squaring old_span_percent in the denominator below allows more aggressive triggering when we are
+    // above desired maximum span and less aggressive triggering when we are far below the desired
+    // maximum span.
+    if ((old_span_percent >= 0.50) && (old_density / old_span_percent_squared < 0.75)) {
+      // We trigger old defragmentation, for example, if:
+      //  old_span_percent is 110% and old_density is below 90.8%, or
+      //  old_span_percent is 100% and old_density is below 75.0%, or
+      //  old_span_percent is  90% and old_density is below 60.8%, or
+      //  old_span_percent is  80% and old_density is below 48.0%, or
+      //  old_span_percent is  70% and old_density is below 36.8%, or
+      //  old_span_percent is  60% and old_density is below 27.0%, or
+      //  old_span_percent is  50% and old_density is below 18.8%.
+      trigger_old_is_fragmented(old_density, first_old_region, last_old_region);
```

Comment:
This method seems a bit useless, why not inline the assignment here? I realize you may be trying to make this method span fewer lines, but I think that can be easily achieved by just moving the description of its work out of line from the code into a block comment at the start that describes how fragmentation is computed and what the trigger threshold criteria are.

Reply:
I'm pulled in both directions on this.  I'll go with your approach, removing both set_fragmentation_trigger() and set_growth_trigger() (renames mentioned in my responses to above comments).